pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,332 @@
+	return imagestreamtagmapper.New(func(in reconcile.Request) []reconcile.Request {
+		if !filter(in.NamespacedName) {
+			return nil
+		}
+		return []reconcile.Request{in}
+	})
+}
+
+type reconciler struct {
+	ctx              context.Context
+	log              *logrus.Entry
+	registryClients  map[string]ctrlruntimeclient.Client
+	pullSecretGetter func() []byte
+}
+
+func (r *reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithField("request", req.String())
+	log.Info("Starting reconciliation")
+	err := r.reconcile(req, log)
+	if err != nil {
+		log.WithError(err).Error("Reconciliation failed")
+	} else {
+		log.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(req reconcile.Request, log *logrus.Entry) error {
+	isTags := map[string]*imagev1.ImageStreamTag{}
+	for clusterName, client := range r.registryClients {
+		*log = *log.WithField("clusterName", clusterName)
+		imageStreamTag := &imagev1.ImageStreamTag{}
+		if err := client.Get(r.ctx, req.NamespacedName, imageStreamTag); err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Debug("Source imageStreamTag not found")
+				continue
+			}
+			return fmt.Errorf("failed to get imageStreamTag %s from registry cluster: %w", req.NamespacedName.String(), err)
+		}
+		isTags[clusterName] = imageStreamTag
+	}
+
+	srcClusterName := findNewest(isTags)
+	if srcClusterName == "" {
+		// the isTag does NOT exist on both clusters
+		// This case is not an error but expected to happen occasionally
+		return nil
+	}
+	sourceImageStreamTag := isTags[srcClusterName]
+
+	imageStreamNameAndTag := strings.Split(req.Name, ":")
+	if n := len(imageStreamNameAndTag); n != 2 {
+		return fmt.Errorf("when splitting imagestreamtagname %s by : expected two results, got %d", req.Name, n)
+	}
+	imageStreamName, imageTag := imageStreamNameAndTag[0], imageStreamNameAndTag[1]
+	isName := types.NamespacedName{Namespace: req.Namespace, Name: imageStreamName}
+	sourceImageStream := &imagev1.ImageStream{}
+	if err := r.registryClients[srcClusterName].Get(r.ctx, isName, sourceImageStream); err != nil {
+		// received a request on the isTag, but the 'is' is no longer there
+		return fmt.Errorf("failed to get imageStream %s from registry cluster: %w", isName.String(), err)
+	}
+
+	*log = *log.WithField("docker_image_reference", sourceImageStreamTag.Image.DockerImageReference)
+
+	for clusterName, client := range r.registryClients {
+		if clusterName == srcClusterName {
+			continue
+		}
+		if err := client.Get(r.ctx, types.NamespacedName{Name: req.Namespace}, &corev1.Namespace{}); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return fmt.Errorf("failed to check if namespace %s exists: %w", req.Namespace, err)
+			}
+			if err := client.Create(r.ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: req.Namespace}}); err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment:
Nit: Can we add a label if we create the NS to identify it as created by this controller?

Reply:
Done.